lib/ex_doc/formatter/html/autolink.ex

```diff
@@ -166,26 +224,34 @@ defmodule ExDoc.Formatter.HTML.Autolink do
         {name, _, args} = form, placeholders when is_atom(name) and is_list(args) ->
           arity = length(args)
 
-          if {name, arity} in typespecs do
-            string = Macro.to_string(form)
-            n = enc_h("#{name}")
-            {string_to_link, _string_with_parens} = split_string_to_link(string)
-            string = ~s[<a href="#t:#{n}/#{arity}">#{h(string_to_link)}</a>]
-
-            put_placeholder(form, string, placeholders)
-          else
-            {form, placeholders}
+          cond do
+            {name, arity} in typespecs ->
+              n = enc_h("#{name}")
+              url = "#t:#{n}/#{arity}"
+              string = format_typespec_form(form, url)
+              put_placeholder(form, string, placeholders)
+
+            {name, arity} in @basic_types ->
+              url = @elixir_docs <> "elixir/" <> @basic_types_page
```

Comment:
TODO: this is not gonna work for hexdocs.pm/elixir/master, v1.5.0, need to find better way of handling this

Reply:
FWIW e.g. EEx _master_ [1] links to Elixir _stable_, so it seems solution above is consistent with that. Then, we'd only need to detect whether we build docs for elixir app itself in which case this would be a relative: `url = @basic_types_page`

WDYT?

https://hexdocs.pm/eex/master/EEx.html#compile_file/2

Reply:
Yeah, I think we need to pipe it though the same logic we link functions today. Then when we change one, everything else works.

Reply:
@josevalim here's another idea: what if we "define" basic & built-in types in the Kernel module?

<img width="858" alt="zrzut ekranu 2017-12-04 o 11 30 30" src="https://user-images.githubusercontent.com/76071/33548160-96138a82-d8e6-11e7-8c98-7d69c11186d0.png">

Another advantage is it magically works from IEx:

<img width="902" alt="zrzut ekranu 2017-12-04 o 11 20 33" src="https://user-images.githubusercontent.com/76071/33547768-7b0416f4-d8e5-11e7-8546-b937e059085e.png">

To make it work, I had to comment out some code in `Kernel.Typespec`:

```elixir
# if builtin_type?(name, arity) do
#   compile_error(caller, "type #{name}/#{arity} is a builtin type and it cannot be redefined")
# end
```

There's similar code in Erlang too, and so I can't yet do:

```elixir
@type term() :: any()
```
One way to solve it is to write it like this instead:

```elixir
@builtin_type type ::
  any()
  | none()
  | ...
  | builtin_type()

@builtin_type builtin_type ::
  term()
  | arity()
  | ...

@basic_type any()

@builtin_type term() :: any()
```

The idea would be that `@basic_type` doesn't actually define the type but makes it accessible by tooling.

Not sure if we could represent https://hexdocs.pm/elixir/typespecs.html#literals section here though.

Reply:
Oh, one other caveat with this: not sure if it'd be technically correct to have types on the Kernel because perhaps they're considered part of the runtime rather.